staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -417,28 +392,351 @@ func TestPriorityAndFairnessWithPanicRecoverAndTimeoutFilter(t *testing.T) {
+		}
+		if response.StatusCode != http.StatusGatewayTimeout {
+			t.Errorf("expected HTTP status code: %d for request: %s, but got: %#v", http.StatusGatewayTimeout, rquestTimesOutPath, response)
+		}
+
+		close(stopCh)
+		t.Log("waiting for the controller to shutdown")
+
+		controllerErr := <-controllerCompletedCh
+		if controllerErr != nil {
+			t.Errorf("expected a nil error from controller, but got: %#v", controllerErr)
+		}
+	})
+
+	t.Run("priority level concurrency is set to 1, inner handler writes to the response before request times out", func(t *testing.T) {
+		const (
+			requestTimeout                                        = 3 * time.Second
+			userName                                              = "alice"
+			fsName                                                = "test-fs"
+			plName                                                = "test-pl"
+			serverConcurrency, plConcurrencyShares, plConcurrency = 1, 1, 1
+		)
+
+		apfConfiguration := newConfiguration(fsName, plName, userName, plConcurrencyShares, 0)
+		stopCh := make(chan struct{})
+		controller, controllerCompletedCh := startAPFController(t, stopCh, apfConfiguration, serverConcurrency, requestTimeout/4, plName, plConcurrency)
+
+		rquestTimesOutPath := "/request/time-out-as-designed"
+		reqHandlerCompletedCh, callerRoundTripDoneCh := make(chan struct{}), make(chan struct{})
+		requestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			expectMatchingAPFHeaders(t, w, fsName, plName)
+
+			if r.URL.Path == rquestTimesOutPath {
+				defer close(reqHandlerCompletedCh)
+
+				// inner handler writes header and then let the request time out.
+				w.WriteHeader(http.StatusBadRequest)
+				<-callerRoundTripDoneCh
+
+				// we expect the timeout handler to have timed out this request by now and any attempt
+				// to write to the response should return a http.ErrHandlerTimeout error.
+				if _, err := w.Write([]byte("foo")); err != http.ErrHandlerTimeout {
+					t.Fatalf("expected error: %#v, but got: %#v", http.ErrHandlerTimeout, err)
+				}
+			}
+		})
+		handler := newHandlerChain(t, requestHandler, controller, userName, requestTimeout)
+
+		server, requestGetter := newHTTP2ServerWithClient(handler)
+		defer server.Close()
+
+		var err error
+		func() {
+			defer close(callerRoundTripDoneCh)
+			t.Logf("waiting for the request: %q to time out", rquestTimesOutPath)
+			_, err = requestGetter(rquestTimesOutPath)
+		}()
+
+		t.Logf("waiting for the inner handler of the request: %q to complete", rquestTimesOutPath)
+		<-reqHandlerCompletedCh
+
+		expectResetStreamError(t, err)
+
+		close(stopCh)
+		t.Log("waiting for the controller to shutdown")
+
+		controllerErr := <-controllerCompletedCh
+		if controllerErr != nil {
+			t.Errorf("expected a nil error from controller, but got: %#v", controllerErr)
+		}
+	})
+
+	t.Run("priority level concurrency is set to 1, queue length is 1, enqueued request should time out with 504", func(t *testing.T) {
```

Comment:
It would be good to describe what is expected for the second request too.  Perhaps something like "priority level concurrency is set to 1, queue length is 1, first request times out, second one does too".